pkg/webhook/server/generic/pod/gracedelete/webhook.go

```diff
@@ -49,42 +49,44 @@ func (gd *GraceDelete) Name() string {
 func (gd *GraceDelete) Validating(ctx context.Context, c client.Client, oldPod, newPod *corev1.Pod, operation admissionv1.Operation) error {
 	// GraceDeleteWebhook FeatureGate defaults to false
 	// Add '--feature-gates=GraceDeleteWebhook=true' to container args, to enable gracedelete webhook
-	if !feature.DefaultFeatureGate.Enabled(features.GraceDeleteWebhook) || operation != admissionv1.Delete {
+	if !feature.DefaultFeatureGate.Enabled(features.GraceDeleteWebhook) || operation != admissionv1.Delete || !utils.ControlledByKusionStack(oldPod) {
 		return nil
 	}
 
-	pod := &corev1.Pod{}
-	if err := c.Get(ctx, types.NamespacedName{Namespace: oldPod.Namespace, Name: oldPod.Name}, pod); err != nil {
-		if !errors.IsNotFound(err) {
-			klog.Error(err, "failed to find pod")
-			return err
-		}
-
-		klog.V(2).Info("pod is deleted")
-		return nil
-	}
-	if !utils.ControlledByKusionStack(pod) {
+	// if pod is allowed to delete
+	if _, allowed := podopslifecycle.AllowOps(poddeletion.OpsLifecycleAdapter, 0, oldPod); allowed {
 		return nil
 	}
 
-	// if Pod is not begin a deletion PodOpsLifecycle, trigger it
-	if !podopslifecycle.IsDuringOps(poddeletion.OpsLifecycleAdapter, pod) {
-		if _, err := podopslifecycle.Begin(c, poddeletion.OpsLifecycleAdapter, pod); err != nil {
-			return fmt.Errorf("fail to begin PodOpsLifecycle to delete Pod %s: %s", pod.Name, err)
+	// label pod to trigger poddeletion_controller reconcile
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		newPod := &corev1.Pod{}
+		err := c.Get(ctx, types.NamespacedName{Namespace: oldPod.Namespace, Name: oldPod.Name}, newPod)
+		if err != nil {
+			return err
 		}
+		if newPod.Labels == nil {
+			newPod.Labels = map[string]string{}
+		}
+		if newPod.Labels[appsv1alpha1.PodDeletionIndicationLabelKey] == "true" {
```

Comment:
The label value would not always be `true`.

Reply:
Row 71 has been fixed as

> 		if _, ok := newPod.Labels[appsv1alpha1.PodDeletionIndicationLabelKey]; ok {
			return nil
		}